### ISSUE TITLE
Default lowering pass for lu_p has accuracy issue.

### ISSUE THREAD

**Issue by ShengYang1** [google/jax#19076]
### Description

The background is, I am developing another device backend for JAX, I found the default lowering pass for lu_p has accuracy issue. It can be reproduced on NV platform.
By default if runnning `python tests/linalg_test.py NumpyLinalgTest.testDetGrad1` it works well.
But if I comment cuda lowering pass [here](https://github.com/google/jax/blob/jaxlib-v0.4.20/jax/_src/lax/linalg.py#L1387-L1390), it goes default pass [here](https://github.com/google/jax/blob/jaxlib-v0.4.20/jax/_src/lax/linalg.py#L1378), then running UT raises error:
```
AssertionError:
Not equal to tolerance rtol=0.1, atol=0.1
JVP primal
x and y nan location mismatch:
 x: array(nan, dtype=float32)
 y: array(0., dtype=float32)
```

### What jax/jaxlib version are you using?

jax 0.4.20, jaxlib 0.4.20+cuda11.cudnn86

### Which accelerator(s) are you using?

GPU

### Additional system info?

python 3.9

**Comment by jakevdp** [comment]
Hi - thanks for the report. It looks like this `_lu_python` implementation is dead code: it is never invoked, because there are platform-specific lowering rules for `cpu`, `cuda`, `rocm`, and `tpu`, which are all the platforms JAX supports.

With that in mind, I think we could probably delete this code, as it is unused. Would that resolve your issue?

**Comment by jakevdp** [comment]
Ah, actually it is not dead code, because it is used in the `jax2tf` lowering path here: https://github.com/google/jax/blob/42ae8432185bf03f61ddd2e7bc279d3abb5247fd/jax/experimental/jax2tf/jax2tf.py#L3299

**Comment by jakevdp** [comment]
The issue is that the `_lu_python` implementation does not handle the gradient of the determinant for zero-valued inputs. Here's a self-contained reproduction of the issue (run on CPU):
```python
import jax.numpy as jnp
from jax._src.interpreters import mlir
from jax.lax import linalg
import jax.test_util as jtu

x = jnp.zeros((3, 3))
assert x.device().platform == 'cpu'
del mlir._platform_specific_lowerings['cpu'][linalg.lu_p]  # this ensures that the default lowering is used.
jtu.check_grads(jnp.linalg.det, (x,), order=1)
```

**Comment by jakevdp** [comment]
Even more minimal:
```python
import jax.numpy as jnp
from jax._src.lax import linalg
x = jnp.zeros((2, 2))

print(linalg.lu(x)[0])
# [[0. 0.]
#  [0. 0.]]

print(linalg._lu_python(x)[0])
# [[ 0.  0.]
#  [nan nan]]
```

**Comment by jakevdp** [comment]
The root cause is the division operation here: https://github.com/google/jax/blob/b07d176419552dd7147da8002aa80fe4b42bc917/jax/_src/lax/linalg.py#L1166-L1168

When `a` is all zeros, then this is `0 / 0` which becomes `NaN`
